### PR TITLE
Typescript

### LINKED ISSUE #70: Move to typescript
Here is what I am thinking so far:

```js

export type EqualityFn = (newArgs: unknown[], lastArgs: unknown[]) => boolean;

function areInputsEqual(
  newInputs: unknown[],
  lastInputs: unknown[],
): boolean {
  // no checks needed if the inputs length has changed
  if (newInputs.length !== lastInputs.length) {
    return false;
  }
  // Using for loop for speed. It generally performs better than array.every
  // https://github.com/alexreardon/memoize-one/pull/59

  for (let i = 0; i < newInputs.length; i++) {
    // using shallow equality check
    if (newInputs[i] !== lastInputs[i]) {
      return false;
    }
  }
  return true;
}


export default function memoizeOne<ResultFn extends (this: unknown, ...newArgs: unknown[]) => ReturnType<ResultFn>>(
  this: unknown,
  resultFn: ResultFn,
  isEqual: EqualityFn = areInputsEqual,
): ResultFn {
  let lastThis: unknown;
  let lastArgs: unknown[] = [];
  let lastResult: ReturnType<ResultFn>;
  let calledOnce: boolean = false;

  // breaking cache when context (this) or arguments change
  function memoized(this: unknown, ...newArgs: unknown[]): ReturnType<ResultFn> {
    if (calledOnce && lastThis === this && isEqual(newArgs, lastArgs)) {
      return lastResult;
    }

    // Throwing during an assignment aborts the assignment: https://codepen.io/alexreardon/pen/RYKoaz
    // Doing the lastResult assignment first so that if it throws
    // nothing will be overwritten
    lastResult = resultFn.apply(this, newArgs);
    calledOnce = true;
    lastThis = this;
    lastArgs = newArgs;
    return lastResult;
  }

  return (memoized as ResultFn);
}

```

[playground](https://www.typescriptlang.org/play/#code/FAUwHgDg9gTgLgAjgTwiBBRAjgVwIYA2AligGIB2CAvAgBTkgDuAgjAOYDOAXAjuQNbkojcgG0AugBoEBPBzitOPPoOFjxASmoA+BACMoUAiDzkA3MGAAzPgGM4RKJTwwQASXIQccDtnwFaYAQEBkYPLx9lASERCUkgmTk4cO9uXmi1OOANHgMjE0oAbwSAehKQqARbAAsQW34OEJAQABNWhCIrJFqOz1SZEHI2OGqEarkq8aHWhM66UJSfADpjIZGEAEIqGll5RY4VweHqrWLg4Nc4HBhKK0IOEAtggF9S8oBVDiIhhCtYGUMEF+-w4aFaSwQbkQbEGIBghAIyAQaBgfxgAFtGnoQHA4HDuqYEC54cgliAAG5w5BvMa4iDcMpsEjVHB6Ja2KDokqEcCuFwtJwldEgdFQIgALxAAFonCASl4CAQSgBWACclmCaLoxkQRGoCAADGYOggADxNMJ9ZarY7GogAantpwSwTKvC+Pw440VwgQIFwhBISJqdX4Lo6XXoTH2oiI4k220SeytHFjmgQZ3OFxx11u90e4deLwSl1zSBgOALr0s4Gg8AQbTuOAIiBs5HsjkowtFEpAAHkGKaAEogDjNuAUP1gPHkFqNWgjIhpFQxcjSJYb0KKZcZWLpqi6EdXG4AFVQIGHo-HFG02kCwUXO9UIni2bHLYoPBH74na9mvgDAgeD8QMyEoGgXHcFMQIIeIcgQb9ryKBIdSTOAT2qJcomfcwUJxNDt2w1cJH1CQngGRBdjgRCWy-HNT3PS8fxvcjUNsBFWgHWwQFyQxjEJGg7gIB4LBpPQ+X4b42CqPAQwQRhakoDlyDxac6EfLR-hcNgcGFFTGhqUwYQSNsOycBBuzFSUWgXTCn1XddNyYQj0hwiR4KPXMzzQJikN0TNgjmWh2MVTj23QAAyCK0IwpdqETR8ECijoAP8KMWHYDhpCo7cNGdLNs2PSgqJouByOLcM3QwmBhCkhtrjqwk5C+NhyD0xA8AMeBGhGdBmqIVr2p4ao6QZEoOTaNByCWRxuWMMA+RgAVyHlQYSiHABNABpKA8HFSrygAETFH5erQ0qiQ4Fq2sGVsiBgeQEA4SoRjwXUuhIboasYDgDoqRcfkYIhFX0dAoEpGBGBgEgZ3DEqrxbfVXGY6a8AgCBEVspdpC3TKNHKmTQpaLj0BoOAKwLLMqNixoybsgmcsy-Vcc4AnSxuc6EbKhJq0KstaEs3sWkuhCuYofHgFeIA)

### PR DESCRIPTION
**Pull request by alexreardon**
Closes #70 

I still need to generate the typescript types for consumers and include a `types` entry into the `package.json` (or figure out what the best practice for this is)

TODO:

- [x] check that things still work for flow consumers
- [x] check that things work for typescript consumers

### CODE REVIEW

File: src/memoize-one.js.flow [hunk 1]
```diff
@@ -0,0 +1,6 @@
+// @flow
+declare module 'memoize-one' {
```
Comment (alexreardon) [inline src/memoize-one.js.flow]: keeping `flow` support. Need to test this with another module consuming it

File: src/memoize-one.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    // Doing the lastResult assignment first so that if it throws
+    // nothing will be overwritten
+    lastResult = resultFn.apply(this, newArgs);
+    calledOnce = true;
+    lastThis = this;
+    lastArgs = newArgs;
+    return lastResult;
+  }
+
+  return memoized as ResultFn;
```
Comment (alexreardon) [inline src/memoize-one.ts]: hard cast back to correct type

**Comment by alexreardon** [comment]
I added a command to generate the typescript declaration files. However, it is creating a module declaration for an internal module `are-inputs-equal`. 

```ts
// We don't want this one - do we!?
declare module "are-inputs-equal" {
    export default function areInputsEqual(newInputs: unknown[], lastInputs: unknown[]): boolean;
}

// The one we actually want
declare module "memoize-one" {
    export type EqualityFn = (newArgs: unknown[], lastArgs: unknown[]) => boolean;
    export function memoizeOne<ResultFn extends (this: any, ...newArgs: any[]) => ReturnType<ResultFn>>(resultFn: ResultFn, isEqual?: EqualityFn): ResultFn;
    export default memoizeOne;
}
```

**Comment by alexreardon** [comment]
I have created this as the `flow` definition file:

```ts
// @flow
declare module 'memoize-one' {
  declare export type EqualityFn = (newArgs: mixed[], lastArgs: mixed[]) => boolean;

  declare export default function memoizeOne<T>(fn: T, isEqual?: EqualityFn): T;
}

```

I then move it to `memoize-one.cjs.js.flow` which `flow` picks up (it is next to `main: memoize-one.cjs.js` in `package.json`).

I am getting the error message for `flow` consumers that I cannot figure out:

> Cannot call memoizeOne because a call signature declaring the expected
parameter / return type is missing in exports

🤷‍♂

**Comment by alexreardon** [comment]
Any thoughts @TrySound? ^

**Comment by PavelVanecekAtlassian** [comment]
Hi Alex! I think the message says _"cannot find the type of this thing"_ in a flowspeak. Following the docs here (and help from MB) makes me believe that the `declare module` syntax is intended to live in `flow-typed` module. Source file like this is supposed to export an actual implementation.

So I think we have two choices here:

Either change the `memoize-one.js.flow` file to be implementation, like this:

```
// @flow

import impl from './impl';

export type EqualityFn = (newArgs: mixed[], lastArgs: mixed[]) => boolean;

const memoizeOne: <T>(fn: T, isEqual?: EqualityFn) => T = impl;

export default memoizeOne;
```

... or distribute the types using https://github.com/flow-typed/flow-typed.

**Comment by alexreardon** [comment]
Thanks @PavelVanecekAtlassian, relying on an implementation worked!

File: rollup.config.js [hunk 3]
```diff
@@ -31,10 +31,10 @@ export default [
       name: 'memoizeOne',
     },
     plugins: [
-      // Setting production env before running babel etc
+      // Setting production env before running other steps
       replace({ 'process.env.NODE_ENV': JSON.stringify('production') }),
-      babel(),
-      commonjs({ include: 'node_modules/**' }),
+      typescript(),
+      // commonjs({ include: 'node_modules/**' }),
```
Comment (alexreardon) [inline rollup.config.js]: I have now removed these comments and the unused dev dep

File: src/memoize-one.ts (hunk 2 above)
Comment (danieldelcore) [reply src/memoize-one.ts]: Hey Alex, how would this look to a consumer?
Do they have to pass the function in explicitly as a generic? 

```typescript
memoizeOne((foo: string, bar: string) => return `${foo}${bar}`);
```

or

```typescript
type Sig = (foo: string, bar: string) => string;
memoizeOne<Sig>((foo: string, bar: string) => return `${foo}${bar}`);
```

File: src/memoize-one.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
+import areInputsEqual from './are-inputs-equal';
+
+export type EqualityFn = (newArgs: unknown[], lastArgs: unknown[]) => boolean;
+
+export default function memoizeOne<
+  // Need to use 'any' rather than 'unknown' here as it has
+  // The correct Generic narrowing behaviour.
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  ResultFn extends (this: any, ...newArgs: any[]) => ReturnType<ResultFn>
```
Comment (danieldelcore) [inline src/memoize-one.ts]: What is `ReturnType`? Is that a TS keyword?

File: src/memoize-one.ts (hunk 2 above)
Comment (alexreardon) [reply src/memoize-one.ts]: Nope, it is inferred 👍

File: rollup.config.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexreardon) [reply rollup.config.js]: they have now been removed. good pick up

File: src/are-inputs-equal.ts [hunk 5]
```diff
@@ -1,8 +1,5 @@
 // @flow
```
[the remark this answers is not in the document]
Comment (alexreardon) [reply src/are-inputs-equal.ts]: 😲

File: src/memoize-one.ts (hunk 4 above)
Comment (alexreardon) [reply src/memoize-one.ts]: It is a typescript utility: https://www.typescriptlang.org/docs/handbook/utility-types.html#returntypet

File: src/are-inputs-equal.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (alexreardon) [reply src/are-inputs-equal.ts]: removed

File: src/memoize-one.ts (hunk 4 above)
Comment (danieldelcore) [reply src/memoize-one.ts]: TIL

File: src/memoize-one.ts (hunk 2 above)
Comment (alexreardon) [reply src/memoize-one.ts]: So the first one: `memoizeOne((foo: string, bar: string) => return `${foo}${bar}`);`

Comment (alexreardon) [reply src/memoize-one.ts]: It pulls the type out 😋

File: rollup.config.js [hunk 6]
```diff
@@ -1,10 +1,9 @@
-// @flow
-import babel from 'rollup-plugin-babel';
-import commonjs from 'rollup-plugin-commonjs';
+// @ts-check
+import typescript from 'rollup-plugin-typescript';
 import replace from 'rollup-plugin-replace';
 import { uglify } from 'rollup-plugin-uglify';
```
Comment (TrySound) [inline rollup.config.js]: Use terser please

File: src/memoize-one.js.flow [hunk 7]
```diff
@@ -0,0 +1,13 @@
+// @flow
+export type EqualityFn = (newArgs: mixed[], lastArgs: mixed[]) => boolean;
+
+export default function memoizeOne<ResultFn: (...any[]) => mixed>(
```
Comment (TrySound) [inline src/memoize-one.js.flow]: `declare` word before allows to omit function body

File: src/memoize-one.js.flow [hunk 8]
```diff
@@ -0,0 +1,13 @@
+// @flow
+export type EqualityFn = (newArgs: mixed[], lastArgs: mixed[]) => boolean;
```
Comment (TrySound) [inline src/memoize-one.js.flow]: Maybe $ReadOnlyArray?

File: src/memoize-one.js.flow (hunk 7 above)
Comment (TrySound) [reply src/memoize-one.js.flow]: does `(...$ReadOnlyArray<empty>) => mixed` work here?

File: test/memoize-one.spec.ts [hunk 9]
```diff
@@ -0,0 +1,727 @@
+import memoize, { EqualityFn } from '../src/memoize-one';
+import isDeepEqual from 'lodash.isequal';
```
Comment (TrySound) [inline test/memoize-one.spec.ts]: I recommend to use `lodash/isEqual` or micro alternative like `dequal`.

File: tsconfig.json [hunk 10]
```diff
@@ -0,0 +1,14 @@
+{
+  "compileOnSave": true,
+  "compilerOptions": {
+    "outDir": "dist",
+    "target": "es5",
+    "module": "commonjs",
```
Comment (TrySound) [inline tsconfig.json]: Does this produce correct output in rollup?

File: test/memoize-one.spec.ts (hunk 9 above)
Comment (alexreardon) [reply test/memoize-one.spec.ts]: this is just for a test. is lodash/isEqual different to lodash.isequal?

File: rollup.config.js (hunk 6 above)
Comment (alexreardon) [reply rollup.config.js]: 👍

File: tsconfig.json (hunk 10 above)
Comment (alexreardon) [reply tsconfig.json]: yeah it seems to. I think the typescript rollup plugin takes care of a few things

File: src/memoize-one.ts (hunk 4 above)
Comment (theKashey) [reply src/memoize-one.ts]: why not ` => any`?

File: test/memoize-one.spec.ts (hunk 9 above)
Comment (TrySound) [reply test/memoize-one.spec.ts]: per method packages will not be supported in lodash v5. Monolithic package is recommended way.

**Pull request merged**